src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -15,102 +15,568 @@
+   * @param id the document id
+   * @param record the record
+   */
+  private void addToRecordMap(String id, SinkRecord record) {
+    if (recordMap != null) {
+      recordMap.put(id, record);
+    }
   }
 
   /**
-   * Gets the Elasticsearch version.
+   * Creates a listener with callback functions to handle completed requests for the BulkProcessor.
    *
-   * @return the version, not null
+   * @return the listener
    */
-  Version getVersion();
+  private BulkProcessor.Listener buildListener() {
+    return new Listener() {
+      @Override
+      public void beforeBulk(long executionId, BulkRequest request) { }
+
+      @Override
+      public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+        for (BulkItemResponse bulkItemResponse : response) {
+          handleResponse(bulkItemResponse);
+          removeFromRecordMap(bulkItemResponse.getId());
+        }
+      }
+
+      @Override
+      public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
+        error.compareAndSet(null, new ConnectException("Bulk request failed.", failure));
+      }
+    };
+  }
 
   /**
-   * Creates indices.
+   * Calls the specified function with retries and backoffs until the retries are exhausted or the
+   * function succeeds.
    *
-   * @param indices the set of index names to create, not null
+   * @param description description of the attempted action n present tense
+   * @param function the function to call and retry
+   * @param <T> the return type of the function
+   * @return the return value of the called function
    */
-  void createIndices(Set<String> indices);
+  private <T> T callWithRetries(String description, Callable<T> function) {
+    try {
+      return RetryUtil.callWithRetries(
+          description,
+          function,
+          config.maxRetries(),
+          config.retryBackoffMs()
+      );
+    } catch (Exception e) {
+      throw new ConnectException("Failed to " + description + ".", e);
+    }
+  }
 
   /**
-   * Creates an explicit mapping.
+   * Configures HTTP authentication and proxy authentication according to the client configuration.
    *
-   * @param index the index to write
-   * @param type the type for which to create the mapping
-   * @param schema the schema used to infer the mapping
-   * @throws IOException if the client cannot execute the request
+   * @param builder the HttpAsyncClientBuilder
    */
-  void createMapping(String index, String type, Schema schema) throws IOException;
+  private void configureAuthentication(HttpAsyncClientBuilder builder) {
+
+    CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+    if (config.isAuthenticatedConnection()) {
+      config.connectionUrls().forEach(
+          url ->
+              credentialsProvider.setCredentials(
+                  new AuthScope(new HttpHost(url)),
+                  new UsernamePasswordCredentials(config.username(), config.password().value())
```

Comment:
Can password be null?

Reply:
nope :), https://github.com/confluentinc/kafka-connect-elasticsearch/blob/master/src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java#L657-L659